### DISCUSSION TITLE
Poor performance of Apple Clang vs. LLVM clang and formulae?

### DISCUSSION THREAD

**Discussion by johnsonjh** [Homebrew/discussions#4361] [Tap maintenance and Homebrew development] [answered]
### Output of `brew config`

```shell
NA
```

### Output of `brew doctor`

```shell
NA
```

### Description of issue

Would it be possible (or even acceptable) to use a specific (non-Xcode/CLT) compiler for building a particular formula for performance reasons?

I’m very familiar with the clever wrappers and processes that Homebrew uses to perform compilations, so that bottles are optimized for the oldest CPU supported by a particular operating system. 

The numbers here are based on optimizing for Core 2 for Intel (so no AVX/AVX2), but benchmarks allowing native optimizations can have an even more drastic performance impact.

In the case of [DPS8M](https://dps8m.gitlab.io/), compilations using the Clang version shipped with Apple’s Xcode CLT 14.2 (or full Xcode 14.3RC2) results in performance that is **30% lower** than then same code built with LLVM 16. For builds from source, allowing AVX2 in both compilers, the Apple Clang performance can be **90% lower** than LLVM 16.

For an M1 machine, an Apple Clang build has **70% lower** performance than an LLVM 16 build.

We are going to release a new version of DPS8M soon, and our own precompiled binaries will be distributing via our homepage will be built using LLVM 16.  While we aren’t even remotely a “popular” package in Homebrew, we’re very happy that Homebrew packages our niche program, but as of the next release, we’ll be updating our documentation to recommend building with LLVM 16 or later.

We maintain [comprehensive documentation](https://dps8m.gitlab.io/dps8m/Documentation/Source_Compilation/) for the recommended compilers for best results on the different systems we support, and it isn’t always straightforward.  There have been cases where a newer compiler on a certain OS has resulted in huge performance regressions, and we’ve worked with the upstreams to resolve those cases.

I also understand that the huge performance gap between LLVM 16 and Apple Clang is likely a temporary situation, but this is perhaps even worse for Homebrew users, because bottles don’t seem to be recompiled due to compiler updates.

So, I have the following questions:

1. Is it allowed to use a particular compiler because of poor performance with Apple’s Clang (or because of the known benefits of the alternative?)

2. Is it allowed to bump a formula - with no other changes - simply to get recompiled bottles when Apple upgrades their compilers, if that’s known to fix these cases?

3. If no to 1 and 2, is it allowed (or appropriate), to mention to users in the caveats text that the binary build installed is known to be drastically disadvantaged?

4. Would it be appropriate for our application to include an OS/compiler “blacklist” which we would use to issue a warning to our users?

5. Some formulas **do** use a particular compiler (LLVM-n, gcc) because they won’t build with Xcode.  Would it be out of bounds for us to intentionally break our build (via ifdef/#error) if the compiler is known to be particularly problematic, to “qualify” us for using a different one.

Now, I really want to recommend on our project webpage that Macintosh users simply use Homebrew to install and stay updated.  However, if the Homebrew builds are hobbled by a 70-90% performance hit compared to build created per our recommendations, we’ll likely end up needing to advise users to avoid using Homebrew builds, at least until the compiler situation resolves.

I do understand we could submit an update to the formula to provide a cask version, or run a tap proving our own binaries, but I think it’s better for everyone to just use Homebrew-core, if possible.

I also want to make it clear that we **do not** want to control the binaries shipped to users, like some “commercial open source” companies do.  In fact, we explicitly and officially support (and document) building with 11 distinct compilers on 14 different operating systems.

Everything is fine *today*, but, we **do** want to avoid the situation where users installing with “brew” have an experience that is drastically poorer than those who don’t, which *is* going to be the case if things remain as they are today, when we make our next release.

Edit: As an aside, this isn’t specific to DPS8M - I noticed that some mathematical software shipped in Homebrew performs abysmally, in fact, essentially unusable. Off the top of my head, Suitesparse is one that seems to be affected, using only a single-thread and even then being considerably slower than one built with a newer Clang version.  On a Mac Studio, it runs *less than one tenth* the speed (that is, **93% slower**) vs. the same source code built with a proper compiler.

**Comment by SMillerDev** [comment, marked as answer]
> Is it allowed to use a particular compiler because of poor performance with Apple’s Clang (or because of the known benefits of the alternative?)

You could make it fail with the default CLang if your project is really so dependent on performance. But generally in the interest of minimising dependencies Homebrew will use system defaults.

> Is it allowed to bump a formula - with no other changes - simply to get recompiled bottles when Apple upgrades their compilers, if that’s known to fix these cases?

If there is a performance fix you can ask a maintainer for a rebuilt. No need to force everyone to a newer version for this I think.

> If no to 1 and 2, is it allowed (or appropriate), to mention to users in the caveats text that the binary build installed is known to be drastically disadvantaged?

As long as it isn't explicitly redirecting away from Homebrew I'd be fine with this.

> Would it be appropriate for our application to include an OS/compiler “blacklist” which we would use to issue a warning to our users?

See 1

> Some formulas do use a particular compiler (LLVM-n, gcc) because they won’t build with Xcode. Would it be out of bounds for us to intentionally break our build (via ifdef/#error) if the compiler is known to be particularly problematic, to “qualify” us for using a different one.

See 1

**Reply by johnsonjh** [reply]
@SMillerDev 

Thanks.  I will figure something out for our next release.

As we are a system simulator (ala QEMU, Hercules, etc.) performance isn’t strictly necessary for functioning, but working with a 50% (or often larger handicap, needlessly)  is something we’d like to avoid.

Hopefully, Apple’s compiler catches up to mainline Clang again.  It’s often been a bit of back and forth between Apple Clang and mainline in regards to performance, but 50-70% difference for all users is too much to ignore. Performance is important to us, but we don’t particularly want to micromanage packaging.  We are also reluctant to add additional maintenance burdens for packagers. I wouldn’t have bothered you all if this was something like a 15% or even 25% difference.

I will also likely make a check and add warning, but in such a way that it isn’t actively steering anyone away from Homebrew specifically.  Homebrew would be preferred!  In fact, we do check (and warn) now if we are running under Rosetta 2 translation, advising the user that performance will be reduced.  This new warning (if macOS && Clang <16) could be handled similarly in the our release.

Thanks.  

I’ll mark this as answered, but, what is the proper way to handle reporting serious performance problems?  There are many.  In the suitesparse case I mentioned above, I see this has [been mentioned here by others](https://github.com/orgs/Homebrew/discussions/4295) too.  This one is really bad, with Brew’s compilation running at only 7% of the expected speed on a real system.  Just open PRs?

**Reply by SMillerDev** [reply]
Just as an additional comment: I think in this case our policy of "upstream should fix it" also applies so I'd definitely be asking for a radr issue for reference if you decide to add it to the formula.

**Comment by carlocab** [comment]
Those are interesting results. Do you get the same performance discrepancy between (non-Apple) LLVM 15 and 16?

I mostly agree with Sean's answers above, for the record. In particular, if you're okay with all your users getting failed builds from trying to use Apple Clang, then option 5 with `#ifdef`/`#error` is fine to me, as long as it results in only a build dependency on LLVM.

**Reply by johnsonjh** [reply]
@carlocab @SMillerDev 

Honestly, I’d prefer just a `#warning` - but many packagers would never see it.  Too bad there isn’t some common convention that would cause CI/automated builds (like BSD ports or Homebrew) to fail, that wouldn’t otherwise be build errors.  Something just to get a maintainer’s attention.

I’ve not checked non-Apple LLVM 15 vs. 16 yet.  I’ll try to find out after lunch.  With DPS8M, even a very simple integer benchmark (n-queens solver time) is representative of the difference:

M1 Mac:
Apple: CPU time 44.520,001,000 - 254,331,969 instructions -> 5.712757 MIPS
LLVM-16: CPU time 28.588,868,000 - 254,331,969 instructions -> 8.896189 MIPS (*55.6918% increase*)

Intel Mac, non-AVX:
Apple: CPU time 53.860,608,000 - 254,331,969 instructions -> 4.722040 MIPS
LLVM-16: CPU time 43.058,378,000 - 254,331,969 instructions -> 5.906678 MIPS (*25% increase*)

With LLVM-16’s Clang, you can see that the Intel x86_64 build actually performs better than the Xcode build running on the M1!  The Apple compiler in both cases was “Apple clang version 14.0.3 (clang-1403.0.22.14.1)”, and the LLVM/Clang compiler was 16.0.0.  All other build variables were the same.

For a bit of background, I have recently proposed a change for our own CI builds to cross-compile from Linux, rather than actually running on MacOS. (Here’s that compiled output, if interested: [dps8m-git-macos-uni.tar.gz](https://github.com/Homebrew/discussions/files/11091000/dps8m-git-macos-uni.tar.gz)). This compile was considerably faster than our old ones built on a Mac.  The build is a 3.0.1-alpha but there shouldn’t have been any changes since 3.0.0 to make it faster.  

A bit of investigation showed that using the newer Clang on the Mac itself also had the same effect, so it thankfully wasn’t something “magical” happening when cross-compiling!

I went right to Clang 16 too (built from source by the [Fedora LLVM team](https://copr.fedorainfracloud.org/coprs/g/fedora-llvm-team/llvm16/), soon to be in F38-Branched) so I didn’t benchmark 15 vs. 16 on Linux either.   

This might be a case of newer Clang just being unusually excellent?

**Reply by johnsonjh** [reply]
As an aside - [we also keep track of performance](https://gitlab.com/dps8m/dps8m/-/wikis/Benchmarks-Historical) and when running on Linux (and most systems), GCC beats Clang by about 12% (on this very specific integer benchmark). Also, GCC has also been improving performance version over version moreso than Clang. We’ll be updating these benchmarks again on the same system for R3.0.1.

Now, macOS is different. GCC builds running on macOS give us terrible performance.  I did a bit of research to see what the difference was, and the extra time the GCC build spends is in “emutls” calls on Mac - which doesn’t happen on Linux.  

It seems this is “emulated TLS” (thread local storage)?  And there isn’t any “non-emulated” TLS in GCC, if I understand correctly.  These results, however, are unsurprising on Mac for us since Clang is (almost) always superior due to Apple’s involvement.

Regardless, I was very surprised by such a huge performance gap between the currently released upstream Clang and the Clang currently shipped by Apple.

**Comment by johnsonjh** [comment]
@carlocab @SMillerDev 

Here’s the results of the some different builds on that same benchmark on an Intel machine.  I ran each test three times, although all runs were within ~1%, and I recorded the best result:

| Package | Compiler | Time | Speed | Diff |
| - | - | - | - | - |
| LLVM 14 | Homebrew Clang 14.0.6 | 47.623550s | 5.340466 MIPS | -1.3% |
| Xcode 14.2 RC2 | Apple Clang 14.0.3 (1403.0.22.14.1) | 47.027942s | 5.408103 MIPS | - |
| LLVM 15 | Homebrew Clang 15.0.7 | 44.600597s | 5.702434 MIPS |  +5.4% |
| LLVM 16 | Homebrew Clang 16.0.0 | 43.058377s | 5.906680 MIPS | + 9.2% |

Clearly, something else must be going wrong (or at least different) with the Homebrew bottle I tested, because repeated runs only got a ~4.9 MIPS score on this benchmark test.  It was reported as being built using Apple Clang 14.0.0.  

I’ll revise, and I now say that the Xcode build performance is not *drastically* slower, however it is *significantly* slower than LLVM 16. The Homebrew bottles, however, are running *drastically* slower, so possibly Apple Clang 14.0.0 vs 14.0.3 is at fault, or maybe another factor?

So, it’s somewhat a mystery.  I’ll continue later tonight once I am home, and I’m going to test all the Homebrew bottles too, and see what my own builds using brew from source yield (these were built with all defaults, except -march=core2 —- scores are about another 10% higher with -march=native allowing for AVX/AVX2).

The pattern here though is still clear, although less drastic than my initial results.  Xcode builds are at least ~10% worse than LLVM 16 builds.  

I’ll now need figure out why the reported (and confirmed) performance of the Homebrew bottle is 20% worse vs LLVM 16 in this test.

Worth noting too, things are almost exactly *twice* as bad if we allowed optimization for Haswell CPUs.  

Tangent: I do understand the rationale for building the most compatible bottles. That said, perhaps it’s time to start offering x86_64h bottles, or even simply move the baseline ISA up?  I know people will complain when their Macs get excluded and obsoleted, but it seems almost unfair that the vast majority of users end up running handicapped binaries?  Haswell CPUs were introduced in Macs a decade ago now.

Now, I’ll be asking my friend with the M1 to run similar tests on his machine again, but all indications are that, so far, LLVM 16 is *significantly* more performant in nearly every case, handily beating Apple Clang (and using the exact same build flags), and usually by about twice the margin as on Intel.

**Comment by johnsonjh** [comment]
Also, these are simple integer tests.  While I can easily come up with specific examples where the differences are an order of magnitude or more pronounced, I’ll stick with these simpler tests to be fair (and not favor vectorization.)

**Comment by johnsonjh** [comment]
Just got results from a Mac Studio (same test):

Apple Clang - 6.832759 MIPS
LLVM-16 Clang - 8.996102 MIPS

So LLVM-16 Clang is a tad bit more than 30% more performant here.
